src/main/java/uk/co/ramp/covid/simulation/population/Population.java

```diff
@@ -51,205 +37,188 @@ public Population(int populationSize, int nHousehold) {
-                        aPopulation[k].setHIndex(i);
-                        this.population[i].addPerson(aPopulation[k]);
-                    }
+    // Checks we have enough people of the right types to perform a household allocation
+    private boolean householdAllocationPossible(BitSet adultIndex, BitSet pensionerIndex,
+                                                BitSet childIndex, BitSet infantIndex) {
+        int adult = 0; int pensioner = 0; int adultPensioner = 0; int adultChild = 0;
+        int pensionerChild = 0; int adultPensionerChild = 0;
+        for (Household h : population) {
+            switch(h.getnType()) {
+                case 1:
+                    adult++;
                     break;
-
-                    case 3:
-                    if (!this.allocationIndex[k] && this.adultIndex[k]) {
-                        this.allocationIndex[k] = true;
-                        aPopulation[k].setHIndex(i);
-                        this.population[i].addPerson(aPopulation[k]);
-                        for (int l = 0; l < this.populationSize; l++) {
-                            if (!this.allocationIndex[l] && this.pensionerIndex[l]) {
-                                this.allocationIndex[l] = true;
-                                aPopulation[k].setHIndex(i);
-                                this.population[i].addPerson(aPopulation[l]);
-                                l = this.populationSize + 1;
-                            }
-                        }
-                    }
+                case 2:
+                    pensioner++;
                     break;
-
-                    case 4:
-                    if (!this.allocationIndex[k] && this.adultIndex[k]) {
-                        this.allocationIndex[k] = true;
-                        aPopulation[k].setHIndex(i);
-                        this.population[i].addPerson(aPopulation[k]);
-                        for (int l = 0; l < this.populationSize; l++) {
-                            if (!this.allocationIndex[l] && this.childIndex[l]) {
-                                this.allocationIndex[l] = true;
-                                aPopulation[k].setHIndex(i);
-                                this.population[i].addPerson(aPopulation[l]);
-                                l = this.populationSize + 1;
-                            }
-                        }
-
-                    }
+                case 3:
+                    adultPensioner++;
                     break;
+                case 4:
+                    adultChild++;
+                    break;
+                case 5:
+                    pensionerChild++;
+                    break;
+                case 6:
+                    adultPensionerChild++;
+                    break;
+            }
+        }
+        return adultIndex.cardinality() <= adult + adultPensioner + adultChild + adultPensionerChild
+                || pensionerIndex.cardinality() <= adultPensioner + pensioner + pensionerChild
```

Comment:
Should this be `&&`, rather than `||`?

Reply:
Good spot! The function name is wrong and should be "householdAllocationImpossible" rather than possible. This way the fast fall-through for || helps us out.

RE timing - I've not got a specific performance test (although we might want to look into a microbenchmarking framework for this). But using just Java nanoTimers for the default population (25k people) the new code is around 200 times faster.

Agree the current test is a sanity check. We might consider adding in features to print out the probability distribution of the households, but checking it in a unit test fashion is difficult due to the randomness in the generation.

Is there a particular part of the logic that was most difficult to follow and I can focus there?
The treatment of infants/children adds quite a bit of additional calls, but hopefully the other calls become clearer once we have the Enums.